python/GafferUI/BoxUI.py

```diff
@@ -149,8 +159,54 @@ def __showContents( graphEditor, box ) :
 
 	GafferUI.GraphEditor.acquire( box )
 
+def __nonDefaultPlugs( box ) :
+
+		result = []
+		def walk( parent ) :
+
+			for p in Gaffer.Plug.Range( parent ) :
+				if p.getName().startswith( "__" ) and parent == box :
+					# Plug won't be exported for referencing, so we
+					# can ignore it.
+					continue
+				if p.getInput() is not None :
+					# Plug will say it is not set to default, but
+					# we don't care as we don't export connections.
+					continue
+				if len( p ) == 0 and isinstance( p, Gaffer.ValuePlug ) :
+					if not p.isSetToDefault() :
```

Comment:
Now it is possible to author true defaults, I think it would be considered bad manners to author a user default. Especially when you consider ExtensionAlgo, you'd be claiming use of a feature which is actually intended for end users of the new node type.

I can see the possibility for confusion though, especially during the transition from the old way. Perhaps this could also be covered by listing the offending plugs in the confirmation dialogue?

Reply:
I guess I was thinking of when tools have been made from Nodes where a facility has used the user defaults mechanism to override the default for some plug on a node at startup.

Reply:
Ah, I see, and that `userDefault` has ended up copied onto the box during plug promotion? I think there are a few possibilities here :

- Don't promote `userDefault`. The mechanism for which metadata to promote is a bit hokey at the moment, as its a list of exclusions passed to `PlugAlgo::promote()` (and hardcoded in practice). I think maybe it'd be better to control the promotion behaviour with metadata. That's similar to what we want to do for the "metadata delegation" thing for presets, so might be best dealt with then.
- Warn if there are user defaults, and provide a separate menu item to remove them. Even if we do the above, this might be useful during the transition.
- Warn if there are user defaults, and remove user defaults automatically in the "Set Default Value" menu items. This seems pretty reasonable, since a user default prevents your authored default from acting the way you want. And it's simpler than a separate menu item and message.

Reply:
I've reworked this to deal with user defaults and to report the details of the errant plugs via the confirmation dialogue in ee6fee4. I'm already unhappy with the verbosity of the message, so haven't made any attempt to distinguish user-default-problems vs not-at-default problems as I think it may just make things more confusing. Not sure if this is the right call or not.

Reply:
Good to see the details there. Certainly helps people know what they might need to attend to. I suggestion for simplifying the text - remove the 'how to' instructions from the main section, and put them at the bottom of the details collapsible (inside, under the list).

Reply:
Good idea - added and squashed in 8706523.